fix: NetworkTransform was not ending extrapolation for some states. [MTT-4521]

NoelStephensUnity opened the pull request:
The most recent `NetworkTransform` update had a regression bug that exposed some potential timing issues with ending extrapolation.  This resolves the issue by migrating the "stop extrapolating the last state" logic over to the non-authoritative side.  This also fixes an issue where axis marked to be synchronized would not continue to interpolate/extrapolate even if the next state did not include deltas.  This also helped resolve the over-all extrapolation issue where axial deltas would start to interpolate towards the target value but could potentially be stopped if another NetworkTransformState was received before the current state was ended.

[MTT-4521](https://jira.unity3d.com/browse/MTT-4521)

## Changelog

- Fixed: Issue where `NetworkTransform` was not ending extrapolation for the previous state causing non-authoritative instances to become out of synch.
- Fixed: Issue where `NetworkTransform` was not continuing to interpolate for the remainder of the associated tick period.
- Fixed: Issue during `NetworkTransform.OnNetworkSpawn` for non-authoritative instances where it was initializing interpolators with the replicated network state which now only contains the transform deltas that occurred during a network tick and not the entire transform state. (WIP-Pending scale synchronization during `NetworkObject` serialization)
- Fixed: Issue where `NetworkTransform` was not honoring the InLocalSpace property on the authority side during OnNetworkSpawn.

## Testing and Documentation
- Includes integration tests:
  - NetworkTransformMultipleChangesOverTime
  - NetworkTransformParentedLocalSpaceTest
- No documentation changes or additions were necessary.

NoelStephensUnity committed:
update

working version that needs clean up

NoelStephensUnity committed:
update

Some touch ups and adjustments to the fix for the extrapolation issue.

NoelStephensUnity committed:
style

adding some minor comments

NoelStephensUnity committed:
update

Added comments and renamed a property.

NoelStephensUnity committed:
update

some code clean up and additional comments

NoelStephensUnity committed:
Merge branch 'develop' into fix/networktransform-extrapolation

NoelStephensUnity committed:
update

MTT-4521
updating changelog file

SamuelBellomo commented on com.unity.netcode.gameobjects/Components/NetworkTransform.cs:
```diff
@@ -910,10 +963,29 @@ private void AddInterpolatedState(NetworkTransformState newState)
+        /// <remarks>
+        /// <see cref="OnNetworkStateChanged"/>
+        /// </remarks>
+        private void StopExtrapolatingLastState()
```
Could prefix name with "Try" as this won't "always" stop the extrapolation

NoelStephensUnity committed:
update

MTT-4521
Renamed StopExtrapolatingLastState to TryToStopExtrapolatingLastState.

Added/adjusted comments.

SamuelBellomo commented on com.unity.netcode.gameobjects/Components/NetworkTransform.cs:
```diff
@@ -910,10 +963,29 @@ private void AddInterpolatedState(NetworkTransformState newState)
+        /// </remarks>
+        private void StopExtrapolatingLastState()
+        {
+            if (!m_LastReceivedState.IsDirty || m_LastReceivedState.EndExtrapolationTick >= NetworkManager.LocalTime.Tick)
```
So does this mean if we haven't received a tick's data for more than a tick, we end the extrapolation? Cause extrapolation is also useful for more than 1 tick. The limit is actually 3x a tick.

NoelStephensUnity replied:
If you look at the original code (line 503) you will see the following:
`else if (!m_HasSentLastValue && m_CachedNetworkManager.LocalTime.Tick >= m_LastSentTick + 1)`
I believe the 3x a tick wiggle room was due to the fact that authority had to send the message to non-authority instances which could take (on average) between 1 and 2 ticks (with the additional tick coming from the m_LastSentTick + 1).
One of the issues with this approach typically involves scenarios where clients with high latencies (128ms+) could take 4+ ticks (TickRate of 30) to just receive the end extrapolation message, which at that point it could be extrapolated beyond the target value(s) which can cause irregular movements on the non-authoritative side.  

What this approach does (handling the end extrapolation 1 tick after the state in question is received) is it removes the extra message (and associated latencies), it assures that it will interpolate up to but not beyond the state's values that changed (yeah it currently does make extrapolation sort of mute).  Then there is a second part of this fix where there is additional code (starting around line 680 the comments describe it in further detail) that keeps interpolating even if the current state has no deltas for an axial value.  This assures that if it does take longer than 1 tick to reach the target value it will still reach the target value (unless a new state is applied that overrides the values being interpolated towards).

Pro's and con's with both approaches...the issue I ran into with additively combining the m_LastSentState with any new state before the end of the current tick and then sending that at the start of a new tick involved the sent time (i.e. when received and adding the measurements it requires the time it was sent...but if it is multiple states combined ( following an "EITHER OR" pattern) then there is just one send state.  So... it was either creating a new message that would send multiple states (with all of that additional overhead) to end extrapolation for each currently in progress state or it was handling it on the non-authoritative side by just creating a copy of a newly received state and applying it a tick later.

Let me see if I can increase the tick offset without impacting anything.

NoelStephensUnity replied:
Yeah... that won't work out properly for this iteration.
We need to revisit how extrapolation is handled while using Quaternion.SlerpUnclamped.

JesseOlmer commented on com.unity.netcode.gameobjects/CHANGELOG.md:
```diff
@@ -10,11 +10,11 @@ Additional documentation and release notes are available at [Multiplayer Documen
 ## [Unreleased]
 
 ### Fixed
-
```
Is there a reason you reordered these? Can you revert the reordering and just include your add?

NoelStephensUnity replied:
We have typically kept the issue # ordered (descending).  The entries were out of order so I ordered them.  
Do you still want me to put them back out of order? (historically we have kept a descending order)

NoelStephensUnity replied:
I could create a separate Jira task for this and put them back into their original non-descending order

JesseOlmer commented on com.unity.netcode.gameobjects/Components/NetworkTransform.cs:
```diff
@@ -420,9 +446,25 @@ internal NetworkVariable<NetworkTransformState> ReplicatedNetworkState
-        private int m_LastSentTick;
+        // Used by both authoritative and non-authoritative instances.
+        // This represents the most recent local authoritative state.
+        private NetworkTransformState m_LocalAuthoritativeNetworkState;
```
Stuff like this is best done in a separate PR from the actual bug fixes / functional changes. It makes the PR harder to read. Can you put it back in its previous place above m_ClientRpcParams?

JesseOlmer commented on com.unity.netcode.gameobjects/Components/NetworkTransform.cs:
```diff
@@ -420,9 +446,25 @@ internal NetworkVariable<NetworkTransformState> ReplicatedNetworkState
+        /// One tick later, NetworkState-A-Post is applied to end that delta's extrapolation.
+        /// <see cref="OnNetworkStateChanged"/> to see how NetworkState-A-Post doesn't get excluded/missed
+        private double m_TickFrequency;
+
```
nit: can you put this above m_LastSentState so that you're not separating that private from its getter method?

JesseOlmer commented on com.unity.netcode.gameobjects/Components/NetworkTransform.cs:
```diff
@@ -1163,8 +1256,31 @@ private void SetStateServerRpc(Vector3 pos, Quaternion rot, Vector3 scale, bool
+        /// If it is not authority it exits early returning false.
+        /// </summary>
+        /// <param name="transformSource">transform to be updated</param>
+        private bool TryUpdateAuthority(Transform transformSource)
```
This name is confusing - it's not a method that Updates the Authority. It's a method that Tries to Commit Transform, and Also Reset Sync State.

I don't fully grok the replicated & local authoritative dirty disparity, but it'd be nice to have a more representative name for this method (or just split It into 2 methods... there's only 2 call sites so it wouldn't be the end of the world)

NoelStephensUnity replied:
Well, I needed to have a single place where authority updates and (to me) it was cleaner this way.
The big reason was that I was starting to have to maintain two sets of code... so perhaps I will just rename it to "UpdateAuthoritativeState", migrate the CanCommitToTransform check back into both methods, and keep the rest of the code/structure.

It doesn't just commit the transform...  :)
It also checks for the NetworkVariable to have been updated, creates a copy of the the local authoritative state, and then resets the local authoritative state.

NoelStephensUnity committed:
update

Applying suggested changes.

NoelStephensUnity committed:
update

removing some changes I made while testing out extrapolating past the target.
adjusted comment.

NoelStephensUnity committed:
style

adjusted comment for clarity purposes.

NoelStephensUnity committed:
test

Adding some of the tests.  Still WIP, but wanted to see how the first completed test fairs in Yamato and on consoles.

NoelStephensUnity committed:
style

removing unused namespace.

NoelStephensUnity committed:
test update

Added a form of tick synchronization prior to making changes to the authoritative transform state.

NoelStephensUnity committed:
style

adjusting comments.

JesseOlmer commented on com.unity.netcode.gameobjects/Components/NetworkTransform.cs:
```diff
@@ -420,14 +450,26 @@ internal NetworkVariable<NetworkTransformState> ReplicatedNetworkState
             return m_LastSentState;
         }
 
+        /// Calculated when spawned, this is used to offset a newly received non-authority side state by 1 tick duration
```
nit: add summary & remarks tags or only use regular // comments

NoelStephensUnity committed:
test

- Added NetworkTransformParentedLocalSpaceTests with associated helper methods, components, and properties.
- Added an "Authority" suffixe to the Authority enums for better clarity when looking at the tests within test runner.
- Updated/added additional comments

JesseOlmer reviewed (approved):
I've approved, but honestly, NetworkTransform is still really hard for me to follow all of the authoritative/non-authoritative/local/interpolator/tick fields and their relationships through this code, so I'm mostly going on faith & our manual test verification here.

NoelStephensUnity committed:
style

Making NetworkTransformParentedLocalSpaceTests singular (NetworkTransformParentedLocalSpaceTest)

NoelStephensUnity committed:
Merge branch 'develop' into fix/networktransform-extrapolation

NoelStephensUnity committed:
style

adding remarks around comments

NoelStephensUnity committed:
style

splitting the comment into a summary with remarks.

ShadauxCat commented on com.unity.netcode.gameobjects/Components/NetworkTransform.cs:
```diff
@@ -666,31 +685,42 @@ private void ApplyAuthoritativeState()
             // InLocalSpace Read:
             InLocalSpace = networkState.InLocalSpace;
 
+            // NOTE ABOUT INTERPOLATING AND BELOW CODE:
```
Thanks for this note. ❤️

ShadauxCat commented on com.unity.netcode.gameobjects/Components/NetworkTransform.cs:
```diff
@@ -706,29 +736,59 @@ private void ApplyAuthoritativeState()
+                interpolatedRotAngles = eulerAngles;
+                if (SyncRotAngleY)
+                {
+                    interpolatedRotAngles.x = m_RotationInterpolator.GetInterpolatedValue().eulerAngles.x;
```
I'm a bit confused here... it looks to me, following up to where interpolatedRotAngles comes from, like you're doing, effectively, `interpolatedRotAngles.x = interpolatedRotAngles.x`. Am I missing something here?

Simplifying what I see:
```C#
interpolatedRotAngles = m_RotationInterpolator.GetInterpolatedValue().eulerAngles;
interpolatedRotAngles.x = m_RotationInterpolator.GetInterpolatedValue().eulerAngles.x;
```

NoelStephensUnity replied:
🤯

NoelStephensUnity replied:
That is a copy and paste bug and I am fixing it post-haste.

NoelStephensUnity committed:
style

updating a comment on the UpdateAuthoritativeState method.

NoelStephensUnity committed:
fix

Copy paste issue.
(need to revisit tests to see why they didn't catch this)

ShadauxCat commented on com.unity.netcode.gameobjects/Components/NetworkTransform.cs:
```diff
@@ -706,29 +736,59 @@ private void ApplyAuthoritativeState()
 
             // Apply the new position
-            if (networkState.HasPositionChange)
+            if (networkState.HasPositionChange || Interpolate && SynchronizePosition)
```
Little bit of a picky comment... I know it's well-defined in the C# standard, but as someone reading the code, operator precedence between `||` and `&&` isn't something I intuitively know... is this 
```C#
(networkState.HasPositionChange || Interpolate) && SynchronizePosition
```
or
```C#
networkState.HasPositionChange || (Interpolate && SynchronizePosition)
```
?

Can you add parentheses to all of these spots where both `||` and `&&` are used to make sure it's clearly communicated (to other developers) which order you're expecting them to be evaluated in?

(This is actually something I'd love if our style could enforced... I've seen a lot of devs - myself included - get bitten by the compiler's order of operations being different than what they were thinking when they wrote the code.)

NoelStephensUnity replied:
I can add that but it will get flagged as "un-needed" parenthesis.
But the general C# Boolean rule is:   && is always handled before ||
(sort of like x + y * 5... where * always happens before +)

ShadauxCat replied:
Huh... that's kind of frustrating, to be honest, that it would flag parentheses added for readability as an issue.

Even knowing the order of operations, I do tend to always write `x + (y * 5)` because the parentheses don't hurt anything and they remove the need to think about the order when reading it - the order's obvious.

NoelStephensUnity replied:
Yeah...I sort of like having parenthesis like that as well...you typically see that more in C++ than C#.
Now, when a parenthesis is needed and won't get flagged:
- `(x + y) * 5`
- `(Something || SomethingElse) && MustBeTrueThing`

ShadauxCat replied:
This preference may indeed be a result of C++ being my native language. :P Most C++ compilers actually flag a warning (at high warning levels) if you DON'T use parentheses in a statement combining `||` and `&&`

NoelStephensUnity replied:
I could be wrong about this...

NoelStephensUnity replied:
I am wrong about this.
It is "ok" because it is an order of operations thing (need to dust off that part of my brain).

> But the general C# Boolean rule is: && is always handled before ||

This is absolutely wrong (my bad).

The thing I was thinking about was how it evaluates the expression (left to right)...Jesse helped clear me up on this.  :)
However, the good news is that one particular section (I evidently was completely not thinking when I typed my response) can have parenthesis without any issues... so I will add those where it is applicable.

NoelStephensUnity committed:
fix

z component

NoelStephensUnity committed:
fix

using the eurlerAngles... that section of code was a mess!
(sorry!)

NoelStephensUnity committed:
update

Just adding some parenthesis per Kitty's suggestion.

NoelStephensUnity committed:
update and fix

Updated ApplyAuthoritativeState so that it is less complicated to follow.
Fixed another issue discovered where the replicated network state values were being used for resetting interpolators during spawn initialization.  Since the replicated network state only has the most recent deltas at the time the NetworkObject is serialized, this would cause "seemingly random" late joining issues.

NoelStephensUnity committed:
test update

This validates the issue with the more recent updates that revealed the network state only having the most recent deltas and not the full transform state.  This resulted in various "seemingly" random late join transform synchronization issues (specifically the scale that is not being synchronized when spawning NetworkObjects).

NoelStephensUnity committed:
update

removed the unnecessary else if (useInterpolatedValue) condition and re-organized it such that it checks to see if it needs to apply interpolation or just depends upon the state to apply the values directly.

Added additional comments about this area of the code and future improvements.

ashwinimurt committed:
Merge branch 'develop' into fix/networktransform-extrapolation

NoelStephensUnity committed:
fix

Removing unclamped slerp and lerp due to a few timing related issues.  These issues should be solved before re-enabling any form of unclamped lerping/slerping with BufferedLinearInterpolator.

Added additional comments in some areas to be investigated for a potential fix.

NoelStephensUnity committed:
fix

temporarily disabling two editor based interpolator extrapolation based tests until I can refactor them to not take extrapolated values into consideration.

NoelStephensUnity committed:
fix

removing debug code used to verify NetworkTransform is not the only thing within NGO that fails when dropping packets (most likely to occur 20-30% drop rate).

NoelStephensUnity committed:
fix

Make sure to initialize based on whether it is set to world or local space.

NoelStephensUnity committed:
Merge branch 'develop' into fix/networktransform-extrapolation

ashwinimurt committed:
Merge branch 'develop' into fix/networktransform-extrapolation

NoelStephensUnity committed:
fix

When authority is applying the current position and rotation during OnNetworkSpawn, use the appropriate local vs world space values.

NoelStephensUnity committed:
update

Adding another fix to the change log

NoelStephensUnity committed:
Merge branch 'develop' into fix/networktransform-extrapolation
